### PR TITLE
[Dataset] Add SuperGlue dataset API

### PR DESCRIPTION
**Pull request by HuXiangkun**
## Description ##
Dataset API of Superglue benchmark as well as unittest. 

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### CODE REVIEW

**Comment by yzh119** [comment]
Related Issue: #795

File: src/gluonnlp/data/dataset.py [hunk 1]
```diff
@@ -296,3 +297,29 @@ def get_field(self, field):
         """
         idx = self._keys.index(field)
         return self._data[idx]
+
+
+class JsonlDataset(SimpleDataset):
```
Comment (eric-haibin-lin) [inline src/gluonnlp/data/dataset.py]: Add doc?

File: src/gluonnlp/data/super_glue.py [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+        super(SuperGlueReCoRD, self).__init__(root, data_file)
+
+    def _read_samples(self, samples):
+        return samples
+
+    def _repo_dir(self):
+        return 'gluon/dataset/SUPERGLUE/ReCoRD'
+
+
+class SuperGlueAX_b(_SuperGlueDataset):
```
Comment (eric-haibin-lin) [inline src/gluonnlp/data/super_glue.py]: SuperGlueAX_b -> SuperGlueAXb

File: src/gluonnlp/data/dataset.py [hunk 3]
```diff
@@ -296,3 +297,38 @@ def get_field(self, field):
...
```
Comment (szha) [inline src/gluonnlp/data/dataset.py]: This is not a concrete class. let's prefix it with underscore `_` and use internally only

**Comment by eric-haibin-lin** [comment]
Could you also add these to the data API list: https://github.com/dmlc/gluon-nlp/blob/master/docs/api/modules/data.rst ?

**Comment by mli** [comment]
Job PR-858/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-858/9/index.html

**Comment by mli** [comment]
Job PR-858/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-858/10/index.html

**Comment by mli** [comment]
Job PR-858/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-858/12/index.html

**Comment by leezu** [comment]
Resolved conflicts.

**Comment by leezu** [comment]
@HuXiangkun 

```
[2020-01-15T16:45:52.943Z] [gw1] [ 20%] FAILED src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueAXb 

[2020-01-15T16:45:53.199Z] src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueAXg 

[2020-01-15T16:45:53.199Z] [gw1] [ 21%] FAILED src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueAXg 

[2020-01-15T16:45:53.467Z] src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueBoolQ 

[2020-01-15T16:45:53.467Z] [gw1] [ 22%] FAILED src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueBoolQ 

[2020-01-15T16:45:53.723Z] src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueCB 

[2020-01-15T16:45:53.723Z] [gw1] [ 24%] FAILED src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueCB 

[2020-01-15T16:45:53.978Z] src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueCOPA 

[2020-01-15T16:45:53.978Z] [gw1] [ 25%] FAILED src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueCOPA 

[2020-01-15T16:45:54.234Z] src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueMultiRC 

[2020-01-15T16:45:54.234Z] [gw1] [ 26%] FAILED src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueMultiRC 

[2020-01-15T16:45:54.234Z] src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueRTE 

[2020-01-15T16:45:54.234Z] [gw1] [ 27%] FAILED src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueRTE 

[2020-01-15T16:45:56.122Z] src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueReCoRD 

[2020-01-15T16:45:56.122Z] [gw1] [ 28%] FAILED src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueReCoRD 

[2020-01-15T16:45:56.378Z] src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueWSC 

[2020-01-15T16:45:56.378Z] [gw1] [ 30%] FAILED src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueWSC 

[2020-01-15T16:45:56.633Z] src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueWiC 

[2020-01-15T16:45:56.633Z] [gw1] [ 31%] FAILED src/gluonnlp/data/super_glue.py::gluonnlp.data.super_glue.SuperGlueWiC 

[2020-01-15T16:45:56.633Z] src/gluonnlp/data/transforms.py::gluonnlp.data.transforms.BERTBasicTokenizer 

```

The doc tests need to be fixed.

**Comment by leezu** [comment]
```
[2020-01-15T16:46:45.615Z] _______________ [doctest] gluonnlp.data.super_glue.SuperGlueAXb ________________

[2020-01-15T16:46:45.615Z] [gw1] linux -- Python 3.5.5 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/conda/cpu/py3/bin/python

[2020-01-15T16:46:45.615Z] 538         Path to temp folder from storing data.

[2020-01-15T16:46:45.615Z] 539         MXNET_HOME defaults to '~/.mxnet'

[2020-01-15T16:46:45.615Z] 540 

[2020-01-15T16:46:45.615Z] 541     Examples

[2020-01-15T16:46:45.615Z] 542     --------

[2020-01-15T16:46:45.615Z] 543     >>> ax_b = gluonnlp.data.SuperGlueAXb(root='./datasets/ax_b')

[2020-01-15T16:46:45.615Z] 544     -etc-

[2020-01-15T16:46:45.615Z] 545     >>> len(ax_b)

[2020-01-15T16:46:45.615Z] 546     1104

[2020-01-15T16:46:45.615Z] 547     >>> ax_b[0].keys()

[2020-01-15T16:46:45.615Z] Expected:

[2020-01-15T16:46:45.615Z]     dict_keys(['idx', 'label', 'sentence1', 'sentence2', 'logic'])

[2020-01-15T16:46:45.615Z] Got:

[2020-01-15T16:46:45.615Z]     dict_keys(['label', 'sentence2', 'sentence1', 'idx', 'logic'])

[2020-01-15T16:46:45.615Z] 

[2020-01-15T16:46:45.615Z] /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/src/gluonnlp/data/super_glue.py:547: DocTestFailure

[2020-01-15T16:46:45.615Z] _______________ [doctest] gluonnlp.data.super_glue.SuperGlueAXg ________________

[2020-01-15T16:46:45.615Z] [gw1] linux -- Python 3.5.5 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/conda/cpu/py3/bin/python

[2020-01-15T16:46:45.615Z] 576         Path to temp folder from storing data.

[2020-01-15T16:46:45.615Z] 577         MXNET_HOME defaults to '~/.mxnet'

[2020-01-15T16:46:45.615Z] 578 

[2020-01-15T16:46:45.615Z] 579     Examples

[2020-01-15T16:46:45.615Z] 580     --------

[2020-01-15T16:46:45.615Z] 581     >>> ax_g = gluonnlp.data.SuperGlueAXg(root='./datasets/ax_g')

[2020-01-15T16:46:45.615Z] 582     -etc-

[2020-01-15T16:46:45.615Z] 583     >>> len(ax_g)

[2020-01-15T16:46:45.615Z] 584     356

[2020-01-15T16:46:45.615Z] 585     >>> ax_g[0].keys()

[2020-01-15T16:46:45.615Z] Expected:

[2020-01-15T16:46:45.615Z]     dict_keys(['hypothesis', 'idx', 'label', 'pair_id', 'premise'])

[2020-01-15T16:46:45.615Z] Got:

[2020-01-15T16:46:45.615Z]     dict_keys(['hypothesis', 'label', 'pair_id', 'idx', 'premise'])

[2020-01-15T16:46:45.615Z] 

[2020-01-15T16:46:45.615Z] /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/src/gluonnlp/data/super_glue.py:585: DocTestFailure

[2020-01-15T16:46:45.615Z] ______________ [doctest] gluonnlp.data.super_glue.SuperGlueBoolQ _______________

[2020-01-15T16:46:45.615Z] [gw1] linux -- Python 3.5.5 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/conda/cpu/py3/bin/python

[2020-01-15T16:46:45.615Z] 437         Path to temp folder from storing data.

[2020-01-15T16:46:45.615Z] 438         MXNET_HOME defaults to '~/.mxnet'

[2020-01-15T16:46:45.615Z] 439 

[2020-01-15T16:46:45.615Z] 440     Examples

[2020-01-15T16:46:45.615Z] 441     --------

[2020-01-15T16:46:45.615Z] 442     >>> boolq_val = gluonnlp.data.SuperGlueBoolQ('val', root='./datasets/boolq')

[2020-01-15T16:46:45.615Z] 443     -etc-

[2020-01-15T16:46:45.615Z] 444     >>> len(boolq_val)

[2020-01-15T16:46:45.615Z] 445     3270

[2020-01-15T16:46:45.615Z] 446     >>> boolq_val[0].keys()

[2020-01-15T16:46:45.615Z] Expected:

[2020-01-15T16:46:45.615Z]     dict_keys(['question', 'passage', 'idx', 'label'])

[2020-01-15T16:46:45.615Z] Got:

[2020-01-15T16:46:45.615Z]     dict_keys(['passage', 'idx', 'question', 'label'])

[2020-01-15T16:46:45.615Z] 

[2020-01-15T16:46:45.615Z] /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/src/gluonnlp/data/super_glue.py:446: DocTestFailure

[2020-01-15T16:46:45.615Z] ________________ [doctest] gluonnlp.data.super_glue.SuperGlueCB ________________

[2020-01-15T16:46:45.615Z] [gw1] linux -- Python 3.5.5 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/conda/cpu/py3/bin/python

[2020-01-15T16:46:45.615Z] 136         Path to temp folder from storing data.

[2020-01-15T16:46:45.615Z] 137         MXNET_HOME defaults to '~/.mxnet'

[2020-01-15T16:46:45.615Z] 138 

[2020-01-15T16:46:45.615Z] 139     Examples

[2020-01-15T16:46:45.615Z] 140     --------

[2020-01-15T16:46:45.615Z] 141     >>> cb_val = gluonnlp.data.SuperGlueCB('val', root='./datasets/cb')

[2020-01-15T16:46:45.615Z] 142     -etc-

[2020-01-15T16:46:45.615Z] 143     >>> len(cb_val)

[2020-01-15T16:46:45.615Z] 144     56

[2020-01-15T16:46:45.615Z] 145     >>> cb_val[0].keys()

[2020-01-15T16:46:45.615Z] Expected:

[2020-01-15T16:46:45.615Z]     dict_keys(['premise', 'hypothesis', 'label', 'idx'])

[2020-01-15T16:46:45.615Z] Got:

[2020-01-15T16:46:45.615Z]     dict_keys(['hypothesis', 'premise', 'idx', 'label'])

[2020-01-15T16:46:45.615Z] 

[2020-01-15T16:46:45.615Z] /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/src/gluonnlp/data/super_glue.py:145: DocTestFailure

[2020-01-15T16:46:45.615Z] _______________ [doctest] gluonnlp.data.super_glue.SuperGlueCOPA _______________

[2020-01-15T16:46:45.615Z] [gw1] linux -- Python 3.5.5 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/conda/cpu/py3/bin/python

[2020-01-15T16:46:45.615Z] 304         Path to temp folder from storing data.

[2020-01-15T16:46:45.615Z] 305         MXNET_HOME defaults to '~/.mxnet'

[2020-01-15T16:46:45.615Z] 306 

[2020-01-15T16:46:45.615Z] 307     Examples

[2020-01-15T16:46:45.615Z] 308     --------

[2020-01-15T16:46:45.615Z] 309     >>> copa_val = gluonnlp.data.SuperGlueCOPA('val', root='./datasets/copa')

[2020-01-15T16:46:45.615Z] 310     -etc-

[2020-01-15T16:46:45.615Z] 311     >>> len(copa_val)

[2020-01-15T16:46:45.615Z] 312     100

[2020-01-15T16:46:45.615Z] 313     >>> copa_val[0].keys()

[2020-01-15T16:46:45.615Z] Expected:

[2020-01-15T16:46:45.615Z]     dict_keys(['premise', 'choice1', 'choice2', 'question', 'label', 'idx'])

[2020-01-15T16:46:45.615Z] Got:

[2020-01-15T16:46:45.615Z]     dict_keys(['label', 'choice1', 'question', 'premise', 'choice2', 'idx'])

[2020-01-15T16:46:45.615Z] 

[2020-01-15T16:46:45.615Z] /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/src/gluonnlp/data/super_glue.py:313: DocTestFailure

[2020-01-15T16:46:45.615Z] _____________ [doctest] gluonnlp.data.super_glue.SuperGlueMultiRC ______________

[2020-01-15T16:46:45.615Z] [gw1] linux -- Python 3.5.5 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/conda/cpu/py3/bin/python

[2020-01-15T16:46:45.615Z] 366     -etc-

[2020-01-15T16:46:45.615Z] 367     >>> len(multirc_val)

[2020-01-15T16:46:45.615Z] 368     83

[2020-01-15T16:46:45.615Z] 369     >>> multirc_val[0].keys()

[2020-01-15T16:46:45.615Z] 370     dict_keys(['text', 'questions'])

[2020-01-15T16:46:45.615Z] 371     >>> len(multirc_val[0]['text'])

[2020-01-15T16:46:45.615Z] 372     12

[2020-01-15T16:46:45.615Z] 373     >>> len(multirc_val[0]['questions'])

[2020-01-15T16:46:45.615Z] 374     13

[2020-01-15T16:46:45.615Z] 375     >>> multirc_val[0]['questions'][0].keys()

[2020-01-15T16:46:45.615Z] Expected:

[2020-01-15T16:46:45.615Z]     dict_keys(['question', 'sentences_used', 'answers', 'idx', 'multisent'])

[2020-01-15T16:46:45.615Z] Got:

[2020-01-15T16:46:45.615Z]     dict_keys(['answers', 'sentences_used', 'idx', 'question', 'multisent'])

[2020-01-15T16:46:45.615Z] 

[2020-01-15T16:46:45.615Z] /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/src/gluonnlp/data/super_glue.py:375: DocTestFailure

[2020-01-15T16:46:45.615Z] _______________ [doctest] gluonnlp.data.super_glue.SuperGlueRTE ________________

[2020-01-15T16:46:45.615Z] [gw1] linux -- Python 3.5.5 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/conda/cpu/py3/bin/python

[2020-01-15T16:46:45.615Z] 081         Path to temp folder for storing data.

[2020-01-15T16:46:45.615Z] 082         MXNET_HOME defaults to '~/.mxnet'.

[2020-01-15T16:46:45.615Z] 083 

[2020-01-15T16:46:45.615Z] 084     Examples

[2020-01-15T16:46:45.615Z] 085     --------

[2020-01-15T16:46:45.615Z] 086     >>> rte_val = gluonnlp.data.SuperGlueRTE('val', root='./datasets/rte')

[2020-01-15T16:46:45.615Z] 087     -etc-

[2020-01-15T16:46:45.615Z] 088     >>> len(rte_val)

[2020-01-15T16:46:45.615Z] 089     277

[2020-01-15T16:46:45.615Z] 090     >>> rte_val[0].keys()

[2020-01-15T16:46:45.615Z] Expected:

[2020-01-15T16:46:45.615Z]     dict_keys(['premise', 'hypothesis', 'label', 'idx'])

[2020-01-15T16:46:45.615Z] Got:

[2020-01-15T16:46:45.615Z]     dict_keys(['hypothesis', 'premise', 'idx', 'label'])

[2020-01-15T16:46:45.615Z] 

[2020-01-15T16:46:45.615Z] /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/src/gluonnlp/data/super_glue.py:90: DocTestFailure

[2020-01-15T16:46:45.615Z] ______________ [doctest] gluonnlp.data.super_glue.SuperGlueReCoRD ______________

[2020-01-15T16:46:45.615Z] [gw1] linux -- Python 3.5.5 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/conda/cpu/py3/bin/python

[2020-01-15T16:46:45.615Z] 489         Path to temp folder from storing data.

[2020-01-15T16:46:45.615Z] 490         MXNET_HOME defaults to '~/.mxnet'

[2020-01-15T16:46:45.615Z] 491 

[2020-01-15T16:46:45.615Z] 492     Examples

[2020-01-15T16:46:45.615Z] 493     --------

[2020-01-15T16:46:45.615Z] 494     >>> record_val = gluonnlp.data.SuperGlueReCoRD('val', root='./datasets/record')

[2020-01-15T16:46:45.615Z] 495     -etc-

[2020-01-15T16:46:45.615Z] 496     >>> len(record_val)

[2020-01-15T16:46:45.615Z] 497     7481

[2020-01-15T16:46:45.615Z] 498     >>> record_val[0].keys()

[2020-01-15T16:46:45.615Z] Expected:

[2020-01-15T16:46:45.615Z]     dict_keys(['source', 'passage', 'qas', 'idx'])

[2020-01-15T16:46:45.615Z] Got:

[2020-01-15T16:46:45.615Z]     dict_keys(['passage', 'source', 'qas', 'idx'])

[2020-01-15T16:46:45.615Z] 

[2020-01-15T16:46:45.615Z] /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/src/gluonnlp/data/super_glue.py:498: DocTestFailure

[2020-01-15T16:46:45.615Z] _______________ [doctest] gluonnlp.data.super_glue.SuperGlueWSC ________________

[2020-01-15T16:46:45.615Z] [gw1] linux -- Python 3.5.5 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/conda/cpu/py3/bin/python

[2020-01-15T16:46:45.615Z] 192         Path to temp folder from storing data.

[2020-01-15T16:46:45.615Z] 193         MXNET_HOME defaults to '~/.mxnet'

[2020-01-15T16:46:45.615Z] 194 

[2020-01-15T16:46:45.615Z] 195     Examples

[2020-01-15T16:46:45.615Z] 196     --------

[2020-01-15T16:46:45.615Z] 197     >>> wsc_val = gluonnlp.data.SuperGlueWSC('val', root='./datasets/wsc')

[2020-01-15T16:46:45.615Z] 198     -etc-

[2020-01-15T16:46:45.615Z] 199     >>> len(wsc_val)

[2020-01-15T16:46:45.615Z] 200     104

[2020-01-15T16:46:45.615Z] 201     >>> wsc_val[0].keys()

[2020-01-15T16:46:45.615Z] Expected:

[2020-01-15T16:46:45.615Z]     dict_keys(['text', 'target', 'idx', 'label'])

[2020-01-15T16:46:45.615Z] Got:

[2020-01-15T16:46:45.615Z]     dict_keys(['text', 'label', 'idx', 'target'])

[2020-01-15T16:46:45.615Z] 

[2020-01-15T16:46:45.615Z] /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/src/gluonnlp/data/super_glue.py:201: DocTestFailure

[2020-01-15T16:46:45.615Z] _______________ [doctest] gluonnlp.data.super_glue.SuperGlueWiC ________________

[2020-01-15T16:46:45.615Z] [gw1] linux -- Python 3.5.5 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/conda/cpu/py3/bin/python

[2020-01-15T16:46:45.615Z] 248         Path to temp folder from storing data.

[2020-01-15T16:46:45.615Z] 249         MXNET_HOME defaults to '~/.mxnet'

[2020-01-15T16:46:45.615Z] 250 

[2020-01-15T16:46:45.615Z] 251     Examples

[2020-01-15T16:46:45.615Z] 252     --------

[2020-01-15T16:46:45.615Z] 253     >>> wic_val = gluonnlp.data.SuperGlueWiC('val', root='./datasets/wic')

[2020-01-15T16:46:45.615Z] 254     -etc-

[2020-01-15T16:46:45.615Z] 255     >>> len(wic_val)

[2020-01-15T16:46:45.616Z] 256     638

[2020-01-15T16:46:45.616Z] 257     >>> wic_val[0].keys()

[2020-01-15T16:46:45.616Z] Expected:

[2020-01-15T16:46:45.616Z]     dict_keys(['word', 'sentence1', 'sentence2', 'idx', 'label', 'start1', 'start2', 'end1', 'end2', 'version'])

[2020-01-15T16:46:45.616Z] Got:

[2020-01-15T16:46:45.616Z]     dict_keys(['label', 'word', 'sentence2', 'end1', 'version', 'start2', 'end2', 'sentence1', 'start1', 'idx'])

[2020-01-15T16:46:45.616Z] 

[2020-01-15T16:46:45.616Z] /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@8/src/gluonnlp/data/super_glue.py:257: DocTestFailure

[2020-01-15T16:46:45.616Z] 

[2020-01-15T16:46:45.616Z] ----------- coverage: platform linux, python 3.5.5-final-0 -----------

```


They are failing now because we decided to support Python 3.5, in which the dictionary order is not well defined.

You need to use `sorted` to make it deterministic.

**Comment by mli** [comment]
Job PR-858/27 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-858/27/index.html

**Comment by szha** [comment]
@leezu thanks for fixing it

**Pull request merged**